File: experiments/006-SmalltalkVsJava/cells/java-gradle_1-Empty_free/2019-2c-parcial2-masked/20260915-111749-40413/output/project/src/main/java/surveycrawler/SurveyCrawler.java
package surveycrawler;

public class SurveyCrawler {

    private Point position;
    private CrawlerFacing facing;
    private final Sonar sonar;
    private CrawlerRun run;

    // instance creation

    public static SurveyCrawler atFacing(Point aPosition, String aFacingName) {
        return atFacingWithSonar(aPosition, aFacingName, new FirmSandSonar());
    }

    public static SurveyCrawler atFacingWithSonar(Point aPosition, String aFacingName, Sonar aSonar) {
        return new SurveyCrawler(aPosition, CrawlerFacing.facing(aFacingName), aSonar);
    }

    public static String invalidFacingErrorDescription() {
        return "Invalid facing";
    }

    // exceptions

    public String invalidCommandErrorDescription() {
        return "Invalid command";
    }

    public String canNotMoveToBoulderErrorDescription() {
        return "Can not move to a boulder";
    }

    public String canNotTurnOnBoulderErrorDescription() {
        return "Can not turn on a boulder";
    }

    public String canNotSlideDuringGuardedRunErrorDescription() {
        return "Can not slide on silt during a guarded run";
    }

    public String canNotStartGuardedRunDuringGuardedRunErrorDescription() {
        return "Can not start a guarded run during a guarded run";
    }

    public String unfinishedGuardedRunErrorDescription() {
        return "Guarded run was not finished";
    }

    public void signalInvalidCommand() {
        throw new RuntimeException(invalidCommandErrorDescription());
    }

    public void signalCanNotMoveToBoulder() {
        throw new RuntimeException(canNotMoveToBoulderErrorDescription());
    }

    public void signalCanNotTurnOnBoulder() {
        throw new RuntimeException(canNotTurnOnBoulderErrorDescription());
    }

    public void signalCanNotSlideDuringGuardedRun() {
        throw new RuntimeException(canNotSlideDuringGuardedRunErrorDescription());
    }

    public void signalCanNotStartGuardedRunDuringGuardedRun() {
        throw new RuntimeException(canNotStartGuardedRunDuringGuardedRunErrorDescription());
    }

    public void signalUnfinishedGuardedRun() {
        throw new RuntimeException(unfinishedGuardedRunErrorDescription());
    }

    // initialization

    private SurveyCrawler(Point aPosition, CrawlerFacing aFacing, Sonar aSonar) {
        position = aPosition;
        facing = aFacing;
        sonar = aSonar;
        run = new UnguardedRun();
    }

    // facing

    public void faceRight() {
        faceTo(new CrawlerFacingRight());
    }

    public void faceUp() {
        faceTo(new CrawlerFacingUp());
    }

    public void faceDown() {
        faceTo(new CrawlerFacingDown());
    }

    public void faceLeft() {
        faceTo(new CrawlerFacingLeft());
    }

    public void faceTo(CrawlerFacing aFacing) {
        facing = aFacing;
    }

    public void turnCounterClockwise() {
        groundAtPosition().turnCrawlerCounterClockwise(this);
    }

    public void turnClockwise() {
        groundAtPosition().turnCrawlerClockwise(this);
    }

    public void turnFacingCounterClockwise() {
        facing.turnCounterClockwise(this);
    }

    public void turnFacingClockwise() {
        facing.turnClockwise(this);
    }

    // testing

    public boolean isAtFacing(Point aPosition, String aFacingName) {
        return position.equals(aPosition) && facing.isFacing(aFacingName);
    }

    public boolean isRetreatCommand(char aCommand) {
        return aCommand == 't';
    }

    public boolean isAdvanceCommand(char aCommand) {
        return aCommand == 'a';
    }

    public boolean isTurnCounterClockwiseCommand(char aCommand) {
        return aCommand == 'g';
    }

    public boolean isTurnClockwiseCommand(char aCommand) {
        return aCommand == 'h';
    }

    // sensing

    private Ground groundAtPosition() {
        return sonar.groundAt(position);
    }

    // moving

    public void retreat() {
        facing.retreat(this);
    }

    public void advance() {
        facing.advance(this);
    }

    public void moveRight() {
        moveTowards(new Point(1, 0));
    }

    public void moveUp() {
        moveTowards(new Point(0, 1));
    }

    public void moveDown() {
        moveTowards(new Point(0, -1));
    }

    public void moveLeft() {
        moveTowards(new Point(-1, 0));
    }

    private void moveTowards(Point aDirection) {
        Point aNewPosition = position.plus(aDirection);
        sonar.groundAt(aNewPosition).moveCrawlerTo(this, aNewPosition, aDirection);
    }

    public void moveTo(Point aNewPosition) {
        run.registerMove(aNewPosition.minus(position));
        position = aNewPosition;
    }

    public void slideTo(Silt aSilt, Point aNewPosition, Point aDirection) {
        run.slideOn(this, aSilt, aNewPosition, aDirection);
    }

    public void undoMove(Point aMove) {
        position = position.minus(aMove);
    }

    // guarded run

    public void startGuardedRun() {
        run.startGuardedRunOn(this);
    }

    public void enterGuardedRun() {
        run = new GuardedRun(facing);
    }

    public void endGuardedRun() {
        run = new UnguardedRun();
    }

    public void undoGuardedRun() {
        run.undoOn(this);
        run = new UnguardedRun();
    }

    // command processing

    public void process(String aSequenceOfCommands) {
        new CommandInterpreter(this, aSequenceOfCommands).process();
    }

    public void processCommand(char aCommand) {
        if (isAdvanceCommand(aCommand)) {
            advance();
            return;
        }
        if (isRetreatCommand(aCommand)) {
            retreat();
            return;
        }
        if (isTurnClockwiseCommand(aCommand)) {
            turnClockwise();
            return;
        }
        if (isTurnCounterClockwiseCommand(aCommand)) {
            turnCounterClockwise();
            return;
        }

        signalInvalidCommand();
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2019-2c-parcial2-masked/20260921-212215-18237/analysis/own-tests-project/src/surveycrawler/survey_crawler.py
from surveycrawler.point import Point
from surveycrawler.crawler_facing import CrawlerFacing
from surveycrawler.crawler_facing_right import CrawlerFacingRight
from surveycrawler.crawler_facing_up import CrawlerFacingUp
from surveycrawler.crawler_facing_down import CrawlerFacingDown
from surveycrawler.crawler_facing_left import CrawlerFacingLeft
from surveycrawler.free_run import FreeRun
from surveycrawler.sonar import Sonar


class SurveyCrawler:

    # instance creation

    @classmethod
    def at_facing(cls, a_position, a_facing_name):
        return cls.at_facing_with_sonar(a_position, a_facing_name, Sonar.all_firm_sand())

    @classmethod
    def at_facing_with_sonar(cls, a_position, a_facing_name, a_sonar):
        return cls(a_position, CrawlerFacing.facing(a_facing_name), a_sonar)

    @classmethod
    def invalid_facing_error_description(cls):
        return "Invalid facing"

    # exceptions

    def invalid_command_error_description(self):
        return "Invalid command"

    def boulder_found_error_description(self):
        return "Can not move to or turn on a boulder"

    def silt_found_error_description(self):
        return "Can not move on silt during a guarded run"

    def guarded_run_inside_guarded_run_error_description(self):
        return "Can not start a guarded run during a guarded run"

    def unfinished_guarded_run_error_description(self):
        return "Guarded run was not finished"

    def signal_invalid_command(self):
        raise RuntimeError(self.invalid_command_error_description())

    def signal_boulder_found(self):
        raise RuntimeError(self.boulder_found_error_description())

    def signal_silt_found(self):
        raise RuntimeError(self.silt_found_error_description())

    def signal_guarded_run_inside_guarded_run(self):
        raise RuntimeError(self.guarded_run_inside_guarded_run_error_description())

    def signal_unfinished_guarded_run(self):
        raise RuntimeError(self.unfinished_guarded_run_error_description())

    # initialization

    def __init__(self, a_position, a_facing, a_sonar):
        self._position = a_position
        self._facing = a_facing
        self._sonar = a_sonar
        self._run = FreeRun()

    # facing

    def face_right(self):
        self.change_facing_to(CrawlerFacingRight())

    def face_up(self):
        self.change_facing_to(CrawlerFacingUp())

    def face_down(self):
        self.change_facing_to(CrawlerFacingDown())

    def face_left(self):
        self.change_facing_to(CrawlerFacingLeft())

    def change_facing_to(self, a_facing):
        # the undo of a turn is turning back to the facing it had before turning
        a_previous_facing = self._facing
        self._facing = a_facing
        self._run.register_undo(lambda: self.change_facing_to(a_previous_facing))

    def turn_counter_clockwise(self):
        self.ground_type_at(self._position).turn_crawler_with(
            self, lambda: self._facing.turn_counter_clockwise(self))

    def turn_clockwise(self):
        self.ground_type_at(self._position).turn_crawler_with(
            self, lambda: self._facing.turn_clockwise(self))

    # testing

    def is_at_facing(self, a_position, a_facing_name):
        return self._position == a_position and self._facing.is_facing(a_facing_name)

    def is_retreat_command(self, a_command):
        return a_command == 't'

    def is_advance_command(self, a_command):
        return a_command == 'a'

    def is_turn_counter_clockwise_command(self, a_command):
        return a_command == 'g'

    def is_turn_clockwise_command(self, a_command):
        return a_command == 'h'

    def is_start_guarded_run_command(self, a_command):
        return a_command == '('

    def is_end_guarded_run_command(self, a_command):
        return a_command == ')'

    def is_repeatable_command(self, a_command):
        return (self.is_advance_command(a_command)
                or self.is_retreat_command(a_command)
                or self.is_turn_clockwise_command(a_command)
                or self.is_turn_counter_clockwise_command(a_command))

    # sounding

    def ground_type_at(self, a_position):
        return self._sonar.ground_type_at(a_position)

    # moving

    def retreat(self):
        self._facing.retreat(self)

    def move_right(self):
        self.move_in_direction(Point(1, 0))

    def advance(self):
        self._facing.advance(self)

    def move_up(self):
        self.move_in_direction(Point(0, 1))

    def move_down(self):
        self.move_in_direction(Point(0, -1))

    def move_left(self):
        self.move_in_direction(Point(-1, 0))

    def move_in_direction(self, a_direction):
        a_new_position = self._position.plus(a_direction)
        self.ground_type_at(a_new_position).move_crawler_to(self, a_new_position, a_direction)

    def slide_to(self, a_position):
        self._run.slide_crawler_to(self, a_position)

    def change_position_to(self, a_position):
        # the undo of a movement is the opposite movement, the one back to where it was
        a_previous_position = self._position
        self._position = a_position
        self._run.register_undo(lambda: self.change_position_to(a_previous_position))

    # guarded run

    def change_to_free_run(self):
        self._run = FreeRun()

    def start_guarded_run(self):
        self._run = self._run.start_guarded_run(self)

    def end_guarded_run(self):
        self._run = self._run.end_guarded_run(self)

    # command processing

    def process(self, a_sequence_of_commands):
        self.change_to_free_run()
        try:
            self.process_commands(a_sequence_of_commands)
        except RuntimeError:
            self._run.handle_error(self)
            raise

    def process_commands(self, a_sequence_of_commands):
        a_last_command = self.no_command()
        for a_command in a_sequence_of_commands:
            a_last_command = self.process_command_or_repetition(a_command, a_last_command)
        self._run.finish(self)

    def no_command(self):
        return ''

    def process_command_or_repetition(self, a_command, a_last_command):
        if a_command.isdigit():
            self.repeat_command_times(a_last_command, int(a_command) + 2)
            return self.no_command()

        self.process_command(a_command)
        return a_command

    def repeat_command_times(self, a_command, a_number_of_times):
        if not self.is_repeatable_command(a_command):
            self.signal_invalid_command()

        for a_repetition in range(a_number_of_times):
            self.process_command(a_command)

    def process_command(self, a_command):
        if self.is_advance_command(a_command):
            self.advance()
            return
        if self.is_retreat_command(a_command):
            self.retreat()
            return
        if self.is_turn_clockwise_command(a_command):
            self.turn_clockwise()
            return
        if self.is_turn_counter_clockwise_command(a_command):
            self.turn_counter_clockwise()
            return
        if self.is_start_guarded_run_command(a_command):
            self.start_guarded_run()
            return
        if self.is_end_guarded_run_command(a_command):
            self.end_guarded_run()
            return

        self.signal_invalid_command()
